biogtr/visualize.py

```diff
@@ -0,0 +1,292 @@
+"""Helper functions for visualizing tracking."""
+from scipy.interpolate import interp1d
+from copy import deepcopy
+from tqdm import tqdm
+from matplotlib import pyplot as plt
+from omegaconf import DictConfig
+
+import seaborn as sns
+import imageio
+import hydra
+import pandas as pd
+import numpy as np
+import cv2
+import imageio
+
+
+palette = sns.color_palette("tab10")
+
+
+def fill_missing(Y: np.ndarray, kind: str = "linear") -> np.ndarray:
+    """Fills missing values independently along each dimension after the first.
+
+    Args:
+        Y: the array for which to fill missing value
+        kind: How to interpolate missing values using `scipy.interpoloate.interp1d`
+
+    Returns:
+        The array with missing values filled in
+    """
+    # Store initial shape.
+    initial_shape = Y.shape
+
+    # Flatten after first dim.
+    Y = Y.reshape((initial_shape[0], -1))
+
+    # Interpolate along each slice.
+    for i in range(Y.shape[-1]):
+        y = Y[:, i]
+
+        # Build interpolant.
+        x = np.flatnonzero(~np.isnan(y))
+        f = interp1d(x, y[x], kind=kind, fill_value=np.nan, bounds_error=False)
+
+        # Fill missing
+        xq = np.flatnonzero(np.isnan(y))
+        y[xq] = f(xq)
+
+        # Fill leading or trailing NaNs with the nearest non-NaN values
+        mask = np.isnan(y)
+        y[mask] = np.interp(np.flatnonzero(mask), np.flatnonzero(~mask), y[~mask])
+
+        # Save slice
+        Y[:, i] = y
+
+    # Restore to initial shape.
+    Y = Y.reshape(initial_shape)
+
+    return Y
+
+
+def annotate_video(
+    video: np.ndarray,
+    labels: pd.DataFrame,
+    key: str,
+    color_palette=palette,
+    trails: bool = True,
+    boxes: int = 64,
+    names: bool = True,
+    centroids: bool = True,
+    poses=False,
+) -> list:
+    """Annotate video frames with labels.
+
+    Labels video with bboxes, centroids, trajectory trails, and/or poses.
+
+    Args:
+        video: The video to be annotated in an ndarray
+        labels: The pandas dataframe containing the centroid and/or pose locations of the instances
```

Comment:
We might want to generalize this - e.g taking predicted instances dict directly. Can handle this once we work on tracker / inference more.